### PR TITLE
Paymaster task

### PR DESCRIPTION
**Pull request by pilagod**

### CODE REVIEW

File: test/utils/AATest.sol [hunk 1]
```diff
@@ -32,6 +33,17 @@ abstract contract AATest is Test {
...
 
+    function handleUserOp(UserOperation memory userOp) internal {
+        UserOperation[] memory ops = new UserOperation[](1);
+        ops[0] = userOp;
+
+        entryPoint.handleOps(ops, payable(msg.sender));
+    }
+
+    function expectRevertFailedOp(string memory reason) internal {
+        vm.expectRevert(abi.encodeWithSelector(IEntryPoint.FailedOp.selector, 0, reason));
```
Comment (NIC619) [inline test/utils/AATest.sol:44]: What does the `0` mean?

File: test/TokenPaymaster.t.sol [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        for (uint8 i = 0; i < logs.length; i++) {
+            if (logs[i].topics[0] == PostOp.selector) {
+                postOp = logs[i];
+                break;
+            }
+        }
+        (, uint256 cost) = abi.decode(postOp.data, (address, uint256));
+        uint256 balance = token.balanceOf(paymaster);
+
+        assertEq(balance, cost);
```
Comment (NIC619) [inline test/TokenPaymaster.t.sol]: Maybe add a comment to explain that this is the `cost` paid by user and the `cost` will go to paymaster?
As it probably does not seem intuitive that cost would equal to balance, usually it's like `post_balance == pre_balance - cost`.

File: test/utils/AATest.sol (hunk 1 above)
Comment (pilagod) [reply test/utils/AATest.sol:44]: First arg for error `FailedOp` is `opIndex`, an indicator for which user op is failed. 

In tests most of the time we would only send one user op and `0` is  enough to cover our current needs. We can overload `expectRevertFailedOp` when we indeed have to check for multiple user ops.

File: test/TokenPaymaster.t.sol (hunk 2 above)
Comment (pilagod) [reply test/TokenPaymaster.t.sol]: Refine in [43b9a00](https://github.com/consenlabs/ethtaipei2023-aa-workshop/pull/8/commits/43b9a0091261a93c9309448a6b98342ccdcf2be5), hope this much clearer 😆

**Review by NIC619 — APPROVED** [review]

File: test/TokenPaymaster.t.sol [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        vm.prank(account);
+        token.approve(paymaster, maxCost);
+
+        userOp.sender = account;
+        userOp.paymasterAndData = abi.encodePacked(paymaster, abi.encode(address(token)));
+
+        vm.recordLogs();
+        handleUserOp(userOp);
+
+        Vm.Log[] memory logs = vm.getRecordedLogs();
```
Comment (108356037) [inline test/TokenPaymaster.t.sol:54]: Nic trick here 👀 !

**Review by 108356037 — APPROVED** [review]

**Pull request merged**